In [ ]:
from sklearn.datasets import load_diabetes
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
# Memuat dataset diabetes
x, y = load_diabetes(return_X_y=True)

In [ ]:
# Inisialisasi model regresi linier
Ir = LinearRegression()

In [ ]:
# Melatih model dengan seluruh data
Ir.fit(x, y)
# Membuat prediksi untuk variabel target berdasarkan data fitur x
y_pred = Ir.predict(x)

In [ ]:
# Menampilkan nilai R^2 untuk mengevaluasi seberapa baik model menjelaskan variabilitas data target
print(r2_score(y, y_pred))

0.5177484222203499


In [ ]:
# Menampilkan nilai Mean Squared Error (MSE) untuk mengukur rata-rata kesalahan kuadrat antara prediksi dan nilai aktual
print(mean_squared_error(y, y_pred))

2859.69634758675


Interpretasi

Nilai R² (R-squared) = 0.5177484222203499,
R² adalah metrik yang menunjukkan seberapa baik model regresi linier dapat menjelaskan variabilitas data.
Nilai 0.517 atau sekitar 51.77% menunjukkan bahwa model dapat menjelaskan sekitar 51.77% dari variabilitas dalam data target.
Ini mengindikasikan bahwa model memiliki kemampuan prediksi yang cukup moderat, karena idealnya nilai R² yang bagus adalah mendekati 1 (100%).

Nilai MSE (Mean Squared Error) = 2859.69634758675,
MSE mengukur rata-rata kesalahan kuadrat antara nilai prediksi dan nilai aktual.
Nilai 2859.69 menunjukkan rata-rata kesalahan prediksi yang cukup besar.
Semakin kecil nilai MSE, semakin akurat model tersebut dalam melakukan prediksi.
Nilai MSE yang besar ini mengindikasikan bahwa prediksi model masih memiliki deviasi yang cukup signifikan dari nilai aktualnya.

In [ ]:
print(x)

[[ 0.03807591  0.05068012  0.06169621 ... -0.00259226  0.01990749
  -0.01764613]
 [-0.00188202 -0.04464164 -0.05147406 ... -0.03949338 -0.06833155
  -0.09220405]
 [ 0.08529891  0.05068012  0.04445121 ... -0.00259226  0.00286131
  -0.02593034]
 ...
 [ 0.04170844  0.05068012 -0.01590626 ... -0.01107952 -0.04688253
   0.01549073]
 [-0.04547248 -0.04464164  0.03906215 ...  0.02655962  0.04452873
  -0.02593034]
 [-0.04547248 -0.04464164 -0.0730303  ... -0.03949338 -0.00422151
   0.00306441]]


In [ ]:
# Membagi data menjadi set pelatihan dan pengujian
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [ ]:
# Menginisialisasi model regresi linier
lr = LinearRegression()
# Melatih model pada data pelatihan
lr.fit(x_train, y_train)
# Membuat prediksi pada data pengujian
y_pred = lr.predict(x_test)
# Menampilkan nilai R^2 untuk mengevaluasi seberapa baik model dalam menjelaskan variabilitas data pengujian
print(r2_score(y_test, y_pred))
# Menampilkan nilai Mean Squared Error (MSE) untuk mengukur rata-rata kesalahan kuadrat antara prediksi dan nilai aktual
print(mean_squared_error(y_test, y_pred))

0.4526027629719195
2900.193628493482


Interpretasi

Nilai R² (R-squared) pada data pengujian = 0.45260276297191953,

Nilai ini lebih rendah dibandingkan R² sebelumnya (yang menggunakan seluruh data).
R² sebesar 0.452 atau sekitar 45.26% menunjukkan bahwa model dapat menjelaskan sekitar 45.26% variabilitas dalam data pengujian.
Penurunan nilai R² dari data keseluruhan (51.77%) ke data pengujian (45.26%) mengindikasikan adanya sedikit overfitting pada model.


Nilai MSE (Mean Squared Error) pada data pengujian = 2900.19362849482,

MSE ini sedikit lebih tinggi dibanding MSE sebelumnya yang menggunakan seluruh data.
Nilai 2900.19 menunjukkan rata-rata kesalahan kuadrat antara prediksi dan nilai aktual pada data pengujian.
Peningkatan nilai MSE dari data keseluruhan (2859.69) ke data pengujian (2900.19) juga mengindikasikan adanya overfitting ringan.
